File: MATLAB Intro/SupplyDemandTax.py
# Code by Javier Gonzalez on October 2019
# For any questions please contact javierj.g18@gmail.com
# Supply and Demand with a tax visualization and wellfare costs.

import numpy as np
import matplotlib.pyplot as plt
from scipy.integrate import quad

T,n= 10, 100
q = np.linspace(0,T,n)

# Ecuations
def p_df(a,b):
    return 15 - b*a
def p_o_wof(a,b):
    return 5 + b*a
def p_o_wf(a,b,c):
    return (5 + b*a)/(1-c)

# Parameters
tau = 0.19
alpha = 1.2

# Store the data
p_o_w = [] #Supply with tax
p_o_wo = [] #Supply without tax
p_d = [] #Demmand

for i in q:
    p_o_wo.append(p_o_wof(i,alpha))
    p_o_w.append(p_o_wf(i,alpha,tau))
    p_d.append(p_df(i,alpha))


#Equilibrium without tax
q_eq_wo = 10/(2*alpha)
p_eq_wo = 15 - alpha*q_eq_wo
integrand = lambda x: 5 + alpha*x
prod_wellfare_wo, error = quad(integrand, 0, q_eq_wo) 
integrand = lambda x: 15 - alpha*x
aux, error = quad(integrand, 0, q_eq_wo) 
consumer_wellfare_wo = aux - prod_wellfare_wo
print("-- Economy Without Taxes --")
print(f"Equilibrium Capital: {q_eq_wo:.2f}")
print(f"Equilibrium Price: {p_eq_wo:.2f}")
print(f"Consumer Wellfare: {consumer_wellfare_wo:.2f}")
print(f"Supply Wellfare: {prod_wellfare_wo:.2f}")
print(f"Total Wellfare: {(prod_wellfare_wo+consumer_wellfare_wo):.2f}")

#Equilibrium with tax
q_eq_w = ((1-tau)*15-5)/(alpha*(2-tau))
p_eq_w = 15 - alpha*q_eq_w
integrand = lambda x: (5 + alpha*x)/(1-tau)
prod_wellfare_w, error = quad(integrand, 0, q_eq_w) 
integrand = lambda x: 15 - alpha*x
aux, error = quad(integrand, 0, q_eq_w) 
consumer_wellfare_w = aux - prod_wellfare_w
fiscal_revenue = p_eq_w*tau*q_eq_w
print("-- Economy With Taxes --")
print(f"Equilibrium Capital: {q_eq_w:.2f}")
print(f"Equilibrium Price: {p_eq_w:.2f}")
print(f"Consumer Wellfare: {consumer_wellfare_w:.2f}")
print(f"Supply Wellfare: {prod_wellfare_w:.2f}")
print(f"Fiscal Revenue: {fiscal_revenue:.2f}")
print(f"Total Wellfare: {(prod_wellfare_w+consumer_wellfare_w+fiscal_revenue):.2f}")

#Change Rates and Loss
print("-- Tax Changes --")
print(f'Consumer Wellfare Change: {(consumer_wellfare_w-consumer_wellfare_wo)/(consumer_wellfare_wo):.2f}%')
print(f'Producers Wellfare Change: {(prod_wellfare_w-prod_wellfare_wo)/(prod_wellfare_wo):.2f}%')
print(f'Total Wellfare Change: {(prod_wellfare_w+consumer_wellfare_w+fiscal_revenue)-(prod_wellfare_wo+consumer_wellfare_wo):.2f}')

# Graph
fig, ax = plt.subplots(figsize=(10, 6))
plt.axis([0, 10,0,18])
ax.plot(q,p_o_wo,'--r' ,label = 'Supply Without Taxes',zorder=-1)
ax.plot(q,p_o_w,'-r', label = 'Supply With Taxes',zorder=-1)
ax.plot(q,p_d, label = 'Demand',zorder=-1)
ax.scatter(q_eq_wo,p_eq_wo,40,'black',zorder=1)
ax.annotate('Equilibrium Without Taxes',xy=(q_eq_wo,p_eq_wo),
            xytext=(q_eq_wo+0.8, p_eq_wo-0.2),fontsize=10,
            arrowprops={'arrowstyle': '<|-'})
ax.scatter(q_eq_w,p_eq_w,40,'black',zorder=1)
ax.annotate('Equilibrium With Taxes',xy=(q_eq_w,p_eq_w),
            xytext=(q_eq_w-1.3, p_eq_w+2),fontsize=10,
            arrowprops={'arrowstyle': '<|-'})
plt.legend()
plt.show()

File: MATLAB Intro/SupplyDemandTax.m
%Code by Javier Gonzalez on October 2019
%For any questions please contact javierj.g18@gmail.com
%Supply and Demand with a tax visualization and wellfare costs.

clear all
%Crear un grid de cantidades
T = 10;
n = 100;
q = transpose(linspace(0,T,n));
%Alocamos la memoria a las matrices
p_o_sin = zeros(n,1);
p_o_con = zeros(n,1);
p_d = zeros(n,1);
%Parametros
tau = 0.19; %Para acercarnos a el IVA
alpha = 1.2;
% Como necesitamos integrar usaremos funciones estaran al final del script
% p_df @(a,alpha) 15 - alpha*a;
% p_o_sinf @(a,alpha) (5 + alpha*a);
% p_o_conf @(a,b,alpha) (5 + alpha*a)/(1-b);


for i = 1:n
    %Llenamos la matriz de la oferta
    p_o_con(i,1) = p_o_conf(q(i,1),alpha,tau);
    %Llenamos oferta sin impuesto
    p_o_sin(i,1) = p_o_sinf(q(i,1),alpha);
    %Llenamos la matriz de la demanda
    p_d(i,1) = p_df(q(i,1),alpha);
end

% Vemos el Grafico
figure
plot(q,p_o_con)
hold on
plot(q,p_o_sin)
plot(q,p_d)
hold off
xlabel('Capital')
legend('Oferta Con Tax','Oferta SIN Tax','Demanda')

%Calculos del equilibrio sin tax
%La cantidad de equilibrio esta dado por:
q_equilibrio_sin = 10/(2*alpha);
% El surplus del productor estara dado por:
prod_wellfare_sin = integral(@(a)p_o_sinf(a,alpha),0,q_equilibrio_sin);
aux = integral(@(a)p_df(a,alpha),0,q_equilibrio_sin);
consumer_wellfare_sin = aux - prod_wellfare_sin;
% Reportamos
disp('-- Economia sin Impuesto --')
z = ['La cantidad de equilibrio es: ', num2str(q_equilibrio_sin)];
disp(z)
z = ['Los productores ganan: ', num2str(prod_wellfare_sin)];
disp(z)
z = ['Los consumidores ganan: ', num2str(consumer_wellfare_sin)];
disp(z)
z = ['WELLFARE TOTAL: ', num2str(consumer_wellfare_sin+prod_wellfare_sin)];
disp(z)

%Calculos del equilibrio con tax
%La cantidad de equilibrio esta dado por:
q_equilibrio_con = ((1-tau)*15-5)/(alpha*(2-tau));
% El surplus del productor estara dado por:
prod_wellfare_con = integral(@(a)p_o_conf(a,alpha,tau),0,q_equilibrio_con);
aux = integral(@(a)p_df(a,alpha),0,q_equilibrio_con);
consumer_wellfare_con = aux - prod_wellfare_con;
%Precio de equilibrio con impuestos es:
impuesto = (15 - alpha*q_equilibrio_con)*(tau);
%La recuadacion del fisco es:
recuadacion = impuesto*q_equilibrio_con;

% Reportamos
disp('-- Economia con Impuesto --')
z = ['La cantidad de equilibrio es: ', num2str(q_equilibrio_con)];
disp(z)
z = ['Los productores ganan: ', num2str(prod_wellfare_con)];
disp(z)
z = ['Los consumidores ganan: ', num2str(consumer_wellfare_con)];
disp(z)
z = ['El fisco recauda: ', num2str(recuadacion)];
disp(z)
z = ['WELLFARE TOTAL: ', num2str(consumer_wellfare_con+prod_wellfare_con+recuadacion)];
disp(z)

% En terminos porcentuales quien pierde mas beneficios?
disp('-- Tasas de Cambio por Impuesto --')
deltacons = (consumer_wellfare_con - consumer_wellfare_sin)/consumer_wellfare_sin;
deltaprod = (prod_wellfare_con - prod_wellfare_sin)/prod_wellfare_sin;
z = ['Cambio del beneficio consumidor con impuesto: ', num2str(deltacons)];
disp(z)
z = ['Cambio del beneficio productor con impuesto: ', num2str(deltaprod)];
disp(z)
z = ['Cambio del beneficio total con impuesto: ', num2str(-(consumer_wellfare_sin+prod_wellfare_sin)+(consumer_wellfare_con+prod_wellfare_con+recuadacion))];
disp(z)

function z = p_df(a,b)
    z = 15 - b*a;
end
function z = p_o_sinf(a,b)
    z = 5 + b*a;
end
function z = p_o_conf(a,b,c)
    z = (5 + b*a)/(1-c);
end
